### ISSUE TITLE
[ENH] Enable PatchTST models within the HuggingFace Forecaster

### ISSUE THREAD

**Issue by benHeid** [sktime/sktime#6709]
Enable the PatchTSMixer architecture within the huggingfaceForecaster. I suppose it should not be that hard. The main issue will be probably that there config differs from the informer/autoformer/etc. models. Thus, there need to be updates made how the configs is build for these models. 

 https://huggingface.co/ibm/patchtsmixer-etth1-generate

**Comment by geetu040** [comment]
@benHeid I think you mixed the 2 models here, although both come from IBM, yet they are different

TSMixer
- Paper: https://arxiv.org/abs/2303.06053
- HF Docs: https://huggingface.co/docs/transformers/en/model_doc/patchtsmixer
- HF Weights: https://huggingface.co/ibm/patchtsmixer-etth1-pretrain

PatchTST
- Paper: https://arxiv.org/pdf/2211.14730
- HF Docs: https://huggingface.co/docs/transformers/en/model_doc/patchtst
- HF Weights: https://huggingface.co/ibm/patchtst-etth1-pretrain

**Comment by Spinachboul** [comment]
@benHeid  @geetu040 
Can I be assigned this issue? I seem to be familiar with `PatchTST` model..

**Comment by Spinachboul** [comment]
@benHeid @geetu040 @fkiraly 
This is the codebase for the same: https://github.com/yuqinie98/PatchTST/
I would propose the following points for implementation:
- The `__init__` should contain hyperparameters like `patch_length`, d_model`, n_heads`.
- We can create a **wrapper class** from the same which will inherit sktime's API.
- There would be a preprocessing step where could create our custom `dataloaders`.
  This also includes creating context points since there would be some sort of mapping between short term semantic 
  info with long term patterns.
- Finally we can use the standard methods `fit()` and `predict` from the `BaseForecaster` class for the training and 
  prediction parts

This is just a high level roadmap but I would request everyone's opinions for the same..
Thanks!!

**Comment by geetu040** [comment]
@Spinachboul Thanks for showing interest and the roadmap.

@julian-fong is working on this here https://github.com/sktime/sktime/pull/6787 by creating a new estimator class `HFPatchTSTForecaster`. We discussed this on discord, if the existing `HFTransformersForecaster` can be extended to use "patch-tst" as well, but there were a few problems and we were looking to generalize the `HFTransformersForecaster` for more than just "patch-tst", so that requires a detailed study of potential models and the current design of `HFTransformersForecaster`.
A few comments for the general design
- `patch_length`, `d_model`, `n_heads` and other such hyperparameters should stay a part of `config`, which is passed through `__init__`
- internally, update the config if needed, on the basis of model architecture, current config changes will fail on other architectures, since they all differ
- task and the relevant class model class depends on the models as well, some use the `ForPrediction` class, some use `ForCausalPrediction`
- during prediction, some models use the `logits` and some rely on the `generate()` method.

**Comment by Spinachboul** [comment]
@geetu040 
That makes a lot of sense!!
Do you think it’d be useful to introduce a common interface or base class for the different transformer-based models, so each one implements certain methods in a consistent way, but the details are left to the model itself?

**Comment by geetu040** [comment]
> Do you think it’d be useful to introduce a common interface or base class for the different transformer-based models, so each one implements certain methods in a consistent way, but the details are left to the model itself?

Yeah, that would be it. So basically we would be switching to a `HFTransformersAdapter`, but I am still not really sure, if we need an adapter or work with the existing forecaster, but yes, that can be an option.

**Comment by Spinachboul** [comment]
@geetu040 
What I was implying to was this concept into practice
Suppose we would want to add `PatchTSTModel` and `BERTModel` , then we would go for something link this
```python
class PatchTSTModel(TransformerModel):
    def __init__(self, config):
        super().__init__(config)
        # parameters specific to model
    
    def predict(self, input_data):
        # Implement the prediction logic specific to PatchTST
        print("Making predictions with PatchTSTModel")
        # return some output
    
    def train(self, training_data):
        # Implement training logic specific to PatchTST
        print("Training PatchTSTModel")
        # return some output
    
    def generate(self, input_data):
        # Implement generation logic if necessary
        print("Generating with PatchTSTModel")
        # return some output


class BERTModel(TransformerModel):
    def __init__(self, config):
        super().__init__(config)
        # parameters specific to model
    
    def predict(self, input_data):
        # Implement the prediction logic specific to BERT
        print("Making predictions with BERTModel")
        # return some output
    
    def train(self, training_data):
        # Implement training logic specific to BERT
        print("Training BERTModel")
        # return some output
    
    def generate(self, input_data):
        # Maybe BERT doesn’t need generate(), or it could use it differently
        print("Generating with BERTModel")
        # return some output
```
Now, the forecaster can use any model that implements the `HFTransformersForecaster` interface, without caring about the specific details of each model.

**Comment by fkiraly** [comment]
closed by https://github.com/sktime/sktime/pull/6787
